### PR TITLE
4966 installer fixes

### PR DESCRIPTION
**Pull request by landreev**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #4966: installer fixes

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [x] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: scripts/installer/install [hunk 1]
```diff
@@ -147,29 +164,75 @@ my $API_URL = "http://localhost:8080/api";
...
 }
 
-# read default configuration values from tab separated file "default.config" if it exists
+# 0c. check if there is the default.config file with the pre-set configuration values: 
+
+#  read default configuration values from tab separated file "default.config" if it exists
 #  moved after the $hostname_from_cmdline section to avoid excessively complicating the logic
 #  of command line argument, automatic selection, or config file.
+#
+#  NOTE: if the file contain any Postgres configuration (for example: "POSTGRES_USER   dvnApp")
```
Comment (matthew-a-dunlap) [inline scripts/installer/install]: Should we note this in our docs?

File: scripts/installer/install [hunk 2]
```diff
@@ -236,72 +306,117 @@ if ($gfdir) {
...
     }
 
-    if ( $yesnocont eq "n" ) {
-        exit 0;
+# 1d. jq executable: 
+
+    my $sys_path = $ENV{'PATH'};
+    my @sys_path_dirs = split( ":", $sys_path );
+
+    if ( $pod_name ne "start-glassfish") # Why is that again?
```
Comment (matthew-a-dunlap) [inline scripts/installer/install]: ?

File: scripts/installer/install [hunk 3]
```diff
@@ -718,7 +771,249 @@ unless ( socket( SOCK, PF_INET, SOCK_STREAM, $rserve_proto )
...
+		print "INFO: /sys/fs/cgroup/memory/memory.limit_in_bytes: $limitline\n";
+		my $limit_in_kb = $limitline / 1024;
+		print "INFO: CGROUP limit_in_kb =  $limit_in_kb [ignoring]\n";
+		# In openshift.json, notice how PostgreSQL and Solr have
+		# resources.limits.memory set to "256Mi".
+		# If you try to give the Dataverse/Glassfish container twice
+		# as much memory (512 MB) and allow $sys_mem_total to
+		# be set below, you should see the following:
+		# "Setting the heap limit for Glassfish to 192MB."
+		# FIXME: dataverse.war will not deploy with only 512 GB of memory.
```
Comment (matthew-a-dunlap) [inline scripts/installer/install]: MB I think

File: scripts/installer/install [hunk 4]
```diff
@@ -871,3 +1134,410 @@ sub set_root_contact_email
...
+	# next line here, just in case: (L.A. -- Sept. 2018)
+	$CONFIG_DEFAULTS{'POSTGRES_ADMIN_PASSWORD'} = $ENV{'POSTGRESQL_ADMIN_PASSWORD'};
+	$CONFIG_DEFAULTS{'POSTGRES_PASSWORD'} = $ENV{"POSTGRES_PASSWORD"};
+    }
+
+    if ( $CONFIG_DEFAULTS{'POSTGRES_SERVER'} eq 'localhost' || $CONFIG_DEFAULTS{'POSTGRES_SERVER'} eq '127.0.0.1' ) 
+    {
+	$pg_local_connection = 1;
+    } 
+#    elsif ($postgresonly)
```
Comment (matthew-a-dunlap) [inline scripts/installer/install]: delete?

**Comment by thaorell** [comment]
Did anyone test this on Openshift or Kubernetes?

**Comment by pdurbin** [comment]
I didn't.

**Comment by kcondon** [comment]
@thaorell , I'll be testing the conventional Dataverse installation scenarios and Danny requested that you test the OpenShift/Kubernetes installation scenario(s). If you have any questions or issues, please check in with Danny. I'm not sure of the timeframe but I hope to finish my testing today or tomorrow.

**Comment by thaorell** [comment]
@kcondon Hi Kevin, I have some work hours tomorrow (11-1) so I will aim to get that done then

**Comment by thaorell** [comment]
Currently, it fails in OpenShift, so I'm assuming it wouldn't work in any containers at all

**Comment by pdurbin** [comment]
Bummer, thanks for testing @thaorell 

Heads up to @kcondon and @landreev ^^

**Comment by landreev** [comment]
@thaorell what part is failing in OpenShift - the postgres setup, or the other part of the installer, or ...? 
Does it have anything to do with how the postgres configuration values are passed to the script via environmental variables under OpenShift? - that was one place I wasn't 100% sure I was doing the right thing. 
But yes, please let us know what it is and we'll figure it out.

**Comment by thaorell** [comment]
@landreev the point of splitting them into 3 scripts are because we want to cut down container startup time. right now, the install script does not run at all for dataverse-glassfish pod. In addition, if we successfully run install with all the postgres user and database creation in dataverse-glassfish application container, it would negate all the efforts to cut down install time.

**Comment by landreev** [comment]
@thaorell there is an option in the installer for skipping the postgres setup (and it should still be skipping it if it sees one of the specified MY_POD_NAME variables - I preserved the checks you put there for that). 
I already mentioned this on slack in response to your question from last night; in dv-install. Actually, it would probably help to keep this conversation in just one place; I don't care if it's here, or in the github issue, or on slack - up to you; but  let's stick to one of these places. 
I understand why you want to do the postgres setup in one place, and the rest of the installer in the other. But I would strongly prefer to keep all the code in the single script, for a million different reasons. (isolating this steps in individual scripts doesn't make them any faster; but it did introduce extra issues and made it harder to maintain).

**Comment by landreev** [comment]
@thaorell 
so this is what the postgres setup section of the script is looking like currently: 
```
# 3. SET UP POSTGRES USER AND DATABASE                                                                                                                                                                   

unless($pod_name eq "start-glassfish" || $pod_name eq "dataverse-glassfish-0" || $skipdatabasesetup) {
    &setup_postgres();
# (there's no return code - if anything goes wrong, the method will exit the script, with some diagnostic messages for the user)                                                                         
    print "\nOK, done.\n";

    if ($postgresonly)
    {
        exit 0;
    }
}
```
meaning - it's skipped, if one of your MY_POD_NAME env. variables is set; or if the --skip_db_setup option has been specified.

**Comment by landreev** [comment]
@thaorell Also, we are for the most part satisfied with how the installer is working for our purposes by now. A couple of extra things came up (completely unrelated to OpenShift) that we'll need to finalize next week; but it should be ready for testing in a container scenario. 
If anything isn't working and/or not working as fast as it was after #4764 - that's unintentional, please let us know what it is and we'll fix it. Once again, the idea was to shuffle some code around, but functionally to keep it working the same exact way as before, for all the OpenShift purposes.

**Comment by landreev** [comment]
@thaorell Hi, sorry if I missed any updates - but what's the current status of deploying under OpenShift from this branch? Is it working or not? - Thanks!

**Pull request merged**